Cython/Utility/FunctionArguments.c

```diff
@@ -250,60 +278,427 @@ static int __Pyx_ParseOptionalKeywords(PyObject *kwds, PyObject *const *kwvalues
+    goto bad;
+invalid_keyword_type:
+    PyErr_Format(PyExc_TypeError,
+        "%.200s() keywords must be strings", function_name);
+    goto bad;
+bad:
+    return -1;
+}
+
+static CYTHON_INLINE int __Pyx_MatchKeywordArg(
+    PyObject *key,
+    PyObject ** const argnames[],
+    PyObject ** const *first_kw_arg,
+    size_t *index_found,
+    const char *function_name)
+{
+    // Optimise for plain str behaviour.
+    return likely(PyUnicode_CheckExact(key)) ?
+        __Pyx_MatchKeywordArg_str(key, argnames, first_kw_arg, index_found, function_name) :
+        __Pyx_MatchKeywordArg_nostr(key, argnames, first_kw_arg, index_found, function_name);
+}
+
+static void __Pyx_RejectUnknownKeyword(
+    PyObject *kwds,
+    PyObject ** const argnames[],
+    PyObject ** const *first_kw_arg,
+    const char *function_name)
+{
+    // Find the first unknown keyword and raise an error. There must be at least one.
+    Py_ssize_t pos = 0;
+    PyObject *key = NULL;
+
+    while (PyDict_Next(kwds, &pos, &key, NULL)) {
+        // Quickly exclude the 'obviously' valid/known keyword arguments (exact pointer match).
+        PyObject** const *name = first_kw_arg;
+        while (*name && (**name != key)) name++;
+
+        if (!*name) {
+            // No exact match found:
+            // compare against positional (always reject) and keyword (reject unknown) names.
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_INCREF(key);
+            #endif
+
+            size_t index_found = 0;
+            int cmp = __Pyx_MatchKeywordArg(key, argnames, first_kw_arg, &index_found, function_name);
+            if (cmp != 1) {
+                if (cmp == 0) {
+                    PyErr_Format(PyExc_TypeError,
+                        "%s() got an unexpected keyword argument '%U'",
+                        function_name, key);
+                }
+                #if CYTHON_AVOID_BORROWED_REFS
+                Py_DECREF(key);
+                #endif
+
+                break;
+            }
+            #if CYTHON_AVOID_BORROWED_REFS
+            Py_DECREF(key);
+            #endif
+        }
+    }
+
+    assert(PyErr_Occurred());
+}
+
+static int __Pyx_ParseKeywordDict(
+    PyObject *kwds,
+    PyObject ** const argnames[],
+    PyObject *values[],
+    Py_ssize_t num_pos_args,
+    const char* function_name,
+    int ignore_unknown_kwargs)
+{
+    PyObject** const *name;
+    PyObject** const *first_kw_arg = argnames + num_pos_args;
+    Py_ssize_t kwcount, extracted = 0;
+    int failed = 0;
+
+    // Check if dict is unicode-keys-only and let Python set the error otherwise.
+    if (unlikely(!PyArg_ValidateKeywordArguments(kwds))) return -1;
+
+    #if CYTHON_COMPILING_IN_CPYTHON && CYTHON_ASSUME_SAFE_SIZE
+    kwcount = PyDict_GET_SIZE(kwds);
+    #else
+    kwcount = PyDict_Size(kwds);
+    if (unlikely(kwcount < 0)) return -1;
+    #endif
+
+    // Extract declared keyword arguments.
+    __Pyx_BEGIN_CRITICAL_SECTION(kwds);
```

Comment:
I think this may be overkill in the sense that Py3.13 is covered by GetItemRef and before that is covered by the GIL

Reply:
I think also both Python and Cython will make sure `kwds` is a fresh uniquely owned dict before the function call. So really the only risk is people manually using the C API to pass a dict which they've also shared with another thread. Which we might able to dismiss as "their fault"